components/psst/psst_rule.cc

```diff
@@ -0,0 +1,160 @@
+
+namespace {
+
+// Format:
+// [
+//   {
+//     "include": [
+//       "*://twitter.com"
+//     ],
+//     "exclude": [
+//     ],
+//     "version": 1,
+//     "test_script": "twitter/test.js",
+//     "policy_script": "twitter/policy.js"
+//   }, ...
+// ]
+// Note that "test_script" and "policy_script" give paths
+// relative to the component under scripts/
+
+// psst.json keys
+const char kInclude[] = "include";
+const char kExclude[] = "exclude";
+const char kTestScript[] = "test_script";
+const char kPolicyScript[] = "policy_script";
+const char kVersion[] = "version";
+
+// Removes trailing dot from |host_piece| if any.
+// Copied from extensions/common/url_pattern.cc
+base::StringPiece CanonicalizeHostForMatching(base::StringPiece host_piece) {
+  if (base::EndsWith(host_piece, ".")) {
+    host_piece.remove_suffix(1);
+  }
+  return host_piece;
+}
+
+}  // namespace
+
+namespace psst {
+
+PsstRule::PsstRule() = default;
+
+PsstRule::~PsstRule() = default;
+
+// static
+bool PsstRule::GetURLPatternSetFromValue(const base::Value* value,
+                                         extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(
+      value->GetList(), URLPattern::SCHEME_HTTP | URLPattern::SCHEME_HTTPS,
+      false, &error);
+  if (!valid) {
+    DVLOG(1) << error;
+  }
+  return valid;
+}
+
+// static
+bool PsstRule::GetFilePathFromValue(const base::Value* value,
+                                    base::FilePath* result) {
+  if (!value->is_string()) {
+    return false;
+  }
+  *result = base::FilePath(value->GetString());
+  return true;
+}
+
+// static
+void PsstRule::RegisterJSONConverter(
+    base::JSONValueConverter<PsstRule>* converter) {
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kInclude, &PsstRule::include_pattern_set_, GetURLPatternSetFromValue);
+  converter->RegisterCustomValueField<extensions::URLPatternSet>(
+      kExclude, &PsstRule::exclude_pattern_set_, GetURLPatternSetFromValue);
+  converter->RegisterCustomValueField<base::FilePath>(
+      kTestScript, &PsstRule::test_script_path_, GetFilePathFromValue);
+  converter->RegisterCustomValueField<base::FilePath>(
+      kPolicyScript, &PsstRule::policy_script_path_, GetFilePathFromValue);
+  converter->RegisterIntField(kVersion, &PsstRule::version_);
+}
+
+// static
+const std::string PsstRule::GetETLDForPsst(const std::string& host) {
+  base::StringPiece host_piece = CanonicalizeHostForMatching(host);
+  return net::registry_controlled_domains::GetDomainAndRegistry(
+      host_piece, net::registry_controlled_domains::PrivateRegistryFilter::
+                      EXCLUDE_PRIVATE_REGISTRIES);
+}
+
+// static
+base::expected<std::pair<std::vector<std::unique_ptr<PsstRule>>,
+                         base::flat_set<std::string>>,
+               std::string>
+PsstRule::ParseRules(const std::string& contents) {
```

Comment:
i assume all rules will be audited and therefore considered "trustworthy", otherwise this needs to follow the rule of 2: https://chromium.googlesource.com/chromium/src/+/master/docs/security/rule-of-2.md

Reply:
Yes the rules will all be audited. Plan is to include the sec team on the [CODEOWNERS file](https://github.com/brave/reviews/issues/1299#issuecomment-1581549933) of the repo we'll eventually set up with all the scripts